pkg/resources/poddistribution/poddistribution.go

```diff
@@ -0,0 +1,386 @@
+	for _, ns := range getNamespaces(nsPrefix, installType) {
+		logrus.Infof("Reconciling Pod Balance in ns %s", ns)
+		unbalanced, err := findUnbalanced(ctx, ns, client)
+		if err != nil {
+			logrus.Errorf("Error getting pods to balance on namespace %s. %v", ns, err)
+		} else {
+			for knn, pods := range unbalanced {
+				objDtls := parseKnn(knn)
+				obj, err := getObject(ctx, client, objDtls)
+				if err != nil || obj == nil {
+					logrus.Errorf("Failed to get object from string %s", knn)
+					continue
+				}
+				balance := verifyBalanceCount(ctx, client, obj)
+				if balance {
+					forceRebalance(ctx, client, objDtls, pods)
+				}
+			}
+		}
+	}
+}
+
+func findUnbalanced(ctx context.Context, nameSpace string, client k8sclient.Client) (map[string][]string, error) {
+	nodes := &corev1.NodeList{}
+	unBalanced := map[string][]string{}
+	if err := client.List(ctx,
+		nodes, &k8sclient.ListOptions{}); err != nil {
+		return unBalanced, err
+	}
+	nodesToZone := map[string]string{}
+	for _, n := range nodes.Items {
+		for _, a := range n.Status.Addresses {
+			if a.Type == corev1.NodeInternalIP {
+				nodesToZone[a.Address] = n.Labels[ZoneLabel]
+				break
+			}
+		}
+	}
+	logrus.Infof("nodes to zone %v", nodesToZone)
+	balance := map[string][]string{}
+	objPods := map[string][]string{}
+	l := &corev1.PodList{}
+	listOpts := []k8sclient.ListOption{
+		k8sclient.InNamespace(nameSpace),
+	}
+	if err := client.List(ctx, l, listOpts...); err != nil {
+		logrus.Errorf("Error getting pod lists %v", err)
+		return unBalanced, err
+	}
+
+	// need to check if there is more than 1 pod of a kind
+	podCount := map[string]int{}
+	logrus.Infof("total pods in ns %s: %d", nameSpace, len(l.Items))
+	for _, p := range l.Items {
+		if p.Status.Phase != "Running" {
+			continue
+		}
+
+		for _, o := range p.OwnerReferences {
+			if *o.Controller {
+				knn := &KindNameSpaceName{
+					NamespacedName: &k8sTypes.NamespacedName{
+						Namespace: nameSpace,
+					},
+				}
+
+				if o.Kind == "ReplicationController" {
+					knn.Name = p.Annotations["openshift.io/deployment-config.name"]
+					knn.Kind = "DeploymentConfig"
+				} else {
+					knn.Name = o.Name
+					knn.Kind = o.Kind
+				}
+				if _, ok := podCount[knn.String()]; !ok {
+					podCount[knn.String()] = 1
+				} else {
+					podCount[knn.String()] = podCount[knn.String()] + 1
+				}
+				if balance[knn.String()] == nil {
+					balance[knn.String()] = []string{}
+				}
+
+				if objPods[knn.String()] == nil {
+					objPods[knn.String()] = []string{}
+				}
+				objPods[knn.String()] = append(objPods[knn.String()], p.Name)
+
+				balance[knn.String()] = append(balance[knn.String()], nodesToZone[p.Status.HostIP])
+				break
+			}
+		}
+	}
+	for knn, z := range balance {
+		if sameZone(z) && podCount[knn] > 1 {
+			logrus.Warningf("Requires pod rebalance %s", knn)
+			unBalanced[knn] = objPods[knn]
+		}
+	}
+
+	return unBalanced, nil
+}
+
+func sameZone(zones []string) bool {
```

Comment:
not sure I get this? If all the pods are in 1 zone will there not just be 1 entry in the array?

Reply:
@maleck13 
No, multiple entries in the array.  example:
DeploymentConfig/redhat-rhoam-3scale/dc1 : ["zone1", "zone1"]

Reply:
removed